lib/datadog/tracing/tracer.rb

```diff
@@ -260,13 +271,32 @@ def continue_trace!(digest, key = nil, &block)
         # Start a new trace from the digest
         context = call_context(key)
         original_trace = active_trace(key)
-        trace = start_trace(continue_from: digest)
+        # When we want the trace to be bound to a block, we cannot let
+        # it auto finish when the local root span finishes. This would
+        # create mutiple traces inside the block. Instead, we'll
+        # expliclity finish the trace after the block finishes.
+        auto_finish = !block
+
+        trace = start_trace(continue_from: digest, auto_finish: auto_finish)
 
         # If block hasn't been given; we need to manually deactivate
         # this trace. Subscribe to the trace finished event to do this.
         subscribe_trace_deactivation!(context, trace, original_trace) unless block
 
-        context.activate!(trace, &block)
+        if block
+          # When a block is given, the trace will be active until the block finishes.
+          context.activate!(trace) do
+            yield
+          ensure # We have to flush even when an error occurs
+            # On block completion, force the trace to finish and flush its finished spans.
+            # Unfinished spans are lost as the {TraceOperation} has ended.
+            trace.finish!
+            flush_trace(trace)
+          end
+        else
+          # Otherwise, the trace will be bound to the current thread after this point
+          context.activate!(trace)
+        end
```

Comment:
This could be streamlined with a guard-clause

```suggestion
        # The trace will be bound to the current thread if no block is given
        return context.activate!(trace) unless block

        # When a block is given, the trace will be active until the block finishes.
        context.activate!(trace) do
          yield
        ensure # We have to flush even when an error occurs
          # On block completion, force the trace to finish and flush its finished spans.
          # Unfinished spans are lost as the {TraceOperation} has ended.
          trace.finish!
          flush_trace(trace)
        end
```